torch/distributed/nn/api/remote_module.py

```diff
@@ -234,18 +238,22 @@ def __init__(
 
             # Create the module on the remote side.
             fut.wait()  # Ensure remote_module_cls is available on remote side.
+            self.module_rref = rpc.rpc_sync(
+                self.on,
+                _create_module_with_interface,
+                (module_cls, args, kwargs, self.device, _module_interface_cls),
+            )
         else:
             self.is_scriptable = False
             self.generated_methods = (
                 _NON_SCRIPTABLE_REMOTE_MODULE_MODULE._generated_methods
             )
-
-        # Create the module on the remote side.
-        self.module_rref = rpc.rpc_sync(
-            self.on,
-            _create_module,
-            (module_cls, args, kwargs, self.device, _module_interface_cls),
-        )
+            # Create the module on the remote side.
+            self.module_rref = rpc.remote(
```

Comment:
Can we do this instead to ensure it is async in all cases:

1. Use `rpc_async` instead of `rpc_sync`, so we now have a Future[RRef] (self.future_rref).
2. Always call `wait()` on self.future_rref in `forward` and `forward_async` and other methods (ex: remote_parameters and get_module_rref) which use `module_rref` before executing the rest of the logic. The `wait()` should be a no-op once the Future has completed.

Reply:
Agree. `rpc_async` should be able to bypass the limitation of `remote` here.

Reply:
@wanchaol Was wondering if you had any ideas of making this async for the torchscript case? Also, see https://github.com/pytorch/pytorch/issues/58098.